R/epichains.R

```diff
@@ -0,0 +1,283 @@
+
+    # print head of the simulation output
+    print(head(x[!is.na(x$ancestor), ]))
+
+    cat("< tree tail >\n")
+
+    # print tail of object
+    print(tail(as.data.frame(x)))
+
+    # print summary information
+    writeLines(
+      c(
+        sprintf("Chains simulated: %s", chain_info[["chains"]]),
+        sprintf(
+          "Unique number of ancestors: %s",
+          chain_info[["unique_ancestors"]]
+        ),
+        sprintf(
+          "Unique number of generations: %s", chain_info[["unique_generations"]]
+        )
+      )
+    )
+
+    # Offer more information to view the full dataset
+    writeLines(sprintf("Use View(<object_name>) to view the full output."))
+
+  } else if (attributes(x)$chain_type == "chains_vec") {
+    cat(sprintf("epichains object \n"))
+    print(as.vector(x))
+    cat(sprintf("Number of chains simulated: %s",
+                chain_info[["unique_chains"]]
+                )
+        )
+    writeLines(
+      c(
+        "\n Simulated chain stats: \n",
+        sprintf("Max: %s", chain_info[["max_chain_stat"]]),
+        sprintf("Min: %s", chain_info[["min_chain_stat"]])
+      )
+    )
+  }
+
+  invisible(x)
+}
+
+
+
+#' Summary method for epichains class
+#'
+#' @param object An [`epichains`] object
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return data frame of information
+#' @export
+summary.epichains <- function(object, ...) {
+  validate_epichains(object)
+
+  if (attributes(object)$chain_type == "chains_tree") {
+
+    chains_ran <- length(object$n)
+
+    max_time <- max(object$time)
+
+    n_unique_ancestors <- length(
+      unique(object$ancestor[!is.na(object$ancestor)])
+    )
+
+    num_generations <- length(unique(object$generation))
+
+    max_generation <- max(object$generation)
+
+    # out of summary
+    res <- list(
+      unique_chains = chains_ran,
+      max_time = max_time,
+      unique_ancestors = n_unique_ancestors,
+      unique_generations = n_unique_ancestors,
+      num_generations = num_generations,
+      max_generation = max_generation
+    )
+  } else if (attributes(object)$chain_type == "chains_vec") {
+    chains_ran <- length(object)
+    max_chain_stat <- max(!is.infinite(object))
+    min_chain_stat <- min(!is.infinite(object))
+
+    res <- list(
+      unique_chains = chains_ran,
+      max_chain_stat = max_chain_stat,
+      min_chain_stat = min_chain_stat
+    )
+  }
+
+  return(res)
+}
+
+#' Checks whether the object is an `epichains`
+#'
+#' @param x An R object
+#'
+#' @return logical, `TRUE` if the object is an `epichains` and `FALSE`
+#' otherwise
+#' @keywords internal
+is_epichains <- function(x) {
+  inherits(x, "epichains")
+}
+
+#' Check if an object is of class "epichains_aggregate_df"
+#'
+#' @param x An [`epichains`] object
+#'
+#' @keywords internal
+is_epichains_aggregate_df <- function(x) {
+  if (!inherits(x, "epichains_aggregate_df")) {
+    stop("Object must have class 'epichains_aggregate_df'")
+  }
+}
```

Comment:
The broader convention for `is_...` functions is to return TRUE / FALSE, not to throw.

Elsewhere, code has `check_...` for the throwing case. Certainly useful to have both, but: generally want to have `check_...` written in terms of `is_` (and per earlier comment, recommend leveraging the type system and using `is.epichains` etc instead), and also exporting `is_...` (definitely something a package consumer might want around).

Reply:
Thanks, Carl. I've modified the `is_` family to return a logical (`TRUE`), and `check_` to throw custom error messages. For example, here d4c6dc9946247a1b06833d8c2d09a9f0080b8a0d. I will implement `is.epichains` method later.

Reply:
Thanks, Carl. Good points there. The `is_` functions are now exported and are used to test for various classes and attributes. The `check_` family will throw errors. Examples of changes are in b99f3b06526bd29f0bac56405e179b5733bb1b44, f7c3db5736020549d1c9eeb993774b51ba4d99ca, d4c6dc9946247a1b06833d8c2d09a9f0080b8a0d.